Webservices

ekcannon opened the pull request:
Full support for REST Web services to expose all BLAST functionality provided by this module.

Commit:
default for blast_threads should be 1

Commit:
Improved text prompts

Commit:
Merge branch '7.x-1.x' of https://github.com/laceysanderson/tripal_blast into 7.x-1.x

Commit:
Fixed URLs for generic links by splitting out any CGI params in URL prefix into array and building link with l()

Commit:
Further fixes to generic linkouts, re: CGI parameters in

Commit:
Fixed sub-parameter problem with GBrowse linkouts

Commit:
Fixed generic links when no CGI parameters are in

Commit:
Moved target dropdown prompt to top of list

prateekgta committed:
Services Deployed

Commit:
Changed NCBI URLs to https

Commit:
Removed unnecessary file

Commit:
Don't remove .pX extension on protein dbs

Commit:
merged

Commit:
name change

Commit:
Name changed

Commit:
Merged changes

Commit:
Merge from upstream repository

Commit:
Removed unecessary system file

Commit:
Removed extra code that was retained incorrectly after merge

Commit:
Restored default GBrowse URL construction

Commit:
Merged changes from upstream repository

Commit:
Added information about setting up and using web services

Commit:
Changed Drupal variable Enable_webservices to enable_blast_webservices

Commit:
Merge branch 'webservices' of https://github.com/LegumeFederation/tripal_blast into webservices

Commit:
The job_id parameter is always all lower case

Commit:
Made getJobId parameters all lower case

Commit:
web service 'getJobId' renamed to 'makeJobRequest'

Commit:
removed unnecessary parameters 'query_type' and 'db_type' from web service getDatabaseOptions

Commit:
Fixed bugs introduced by changes to getDatabaseOptions service

Commit:
Updated and corrected web services documentation

Commit:
Re-fixed parameter for getDatabaseOption

Commit:
Included minimum BLAST+ version

njbooher commented:
run_BLAST_tripal_job_services in this pull request has the same security problems that run_BLAST_tripal_job did before 434279e7a45e32c66624862589909e4146292344 . Please do not mege this without fixing them.

laceysanderson commented:
Thanks for the comment @njbooher, I see that and will make sure :-)

laceysanderson commented:
Ethy, a lot of the conflict here is because  our repos have diverged quite a bit. I suggest we start by resolving the main repo 7.x-1.x branches so that you can take advantage of the security and functionality fixes we've been hard at work on lately (See #17). Once we get that resolved then you can pull the changes from your main branch into your web services and it should make this pull request quite a bit easier. Sound like a plan?

laceysanderson commented on api/blast_ui.api.inc:
```diff
@@ -1,12 +1,91 @@
+  $blast_formatter_command = $blast_path . 'blast_formatter';
+
+  //$blast_cmd = "$program -query '$query' -db '$database' -out '$output_file' -outfmt=11";
+  $blast_cmd = "$program -query '$query' -db '$database' -out '$output_file' -outfmt=".$format;
```
We have to be very careful at this point to ensure we don't bring in a security risk. When generating the blast command we have to make sure to escapeshellarg() every single parameter as we add it to the command. This ensures that if our validation fails at some point prior we don't end up with arbitrary code executed. See https://github.com/tripal/tripal_blast/blob/7.x-1.x/api/blast_ui.api.inc#L215 for an example.

Commit:
Merged with main repository

Commit:
Giving credit where credit is due

Commit:
Added escapeshellarg to command construction

Commit:
Changed hard-coded temp directory to use Drupal's temp directory

laceysanderson commented on api/blast_ui.api.inc:
```diff
@@ -1,12 +1,96 @@
+
+function run_BLAST_tripal_job_services($program, $query, $database, $output_filestub, $options, $format, $format_type,  $job_id = NULL) {
+  $output_file = file_directory_temp() .  DIRECTORY_SEPARATOR . $output_filestub . '.blast.' . $format_type;
+  $output_dir = variable_get('file_public_path', conf_path() . '/files') 
```
$output_file is using the tmp directory for the results file when this module using the files directory (as specified in $output_dir) for results. Results should be put in the same place regardless of whether they were run via web services or through the GUI. 

`$output_file = $output_dir . DIRECTORY_SEPARATOR . $output_filestub . '.blast.' . $format_type;`

laceysanderson commented on api/blast_ui.api.inc:
```diff
@@ -1,12 +1,96 @@
+ * @param $format 
+ *   Which specifies Json or asn format of file or program
+ * @param  $format_type 
+ * which specify the number of format used by BLAST e.g json has 11 and json as 12. 
```
Documentation for $format and $format_type needs to be improved. Are you expecting $format = 'JSON' or $format = 12? Same goes for $format_type. Also, you say that json is both 11 and 12 (assuming this is a typo).

laceysanderson commented on api/blast_ui.api.inc:
```diff
@@ -1,12 +1,96 @@
+  $blast_threads = variable_get('blast_threads', '');
+  
+  // Strip the extension off the BLAST target
+  $database = preg_replace("/(.*)\.[pn]\w\w/", '$1', $database);
```
We need to be consistent with how we are handling this. Here you have the old code (very buggy) and in the original run blast function you've commented it out entirely. I suggest we stick with my most recent fix of changing the regex to ensure the "file extension" captured is at the end and only 3 characters for now since. If you feel strongly that this issue should be dealt with immediately, I don't want to comment it out until validation is added to the admin form to node form to ensure that admin don't enter the extension in the first place.

ekcannon replied:
All instances of the regex's have been removed and a check added to the node validation to ensure a file type extension was not added to the BLAST target path and file prefix.

laceysanderson commented on api/blast_ui.api.inc:
```diff
@@ -84,12 +169,16 @@ function get_blast_database_options($type) {
-  $options[0] = 'Select a Dataset';
+  
+  // Prepend prompt; note that array_unshift() will re-index the array
+  $options = array('Select a Dataset') + $options;
```
It looks like we should be adding the empty option when we define the form element rather then here. Specifically, by adding `#empty_option` to https://github.com/tripal/tripal_blast/blob/7.x-1.x/includes/blast_ui.form_per_program.inc#L225 like this:

```
  $options = get_blast_database_options($db_type);
  $form['B']['DB']['SELECT_DB'] = array(
    '#type' => 'select',
    '#title' => t('%type BLAST Databases:', array('%type' => ucfirst($db_type))),
    '#options' => $options,
    '#empty_option' => t('- Select a Dataset -'),
    '#default_value' => $defaults['SELECT_DB'],
  );
```

ekcannon replied:
Makes sense. Keeps get_blast_database_options() more general too.

laceysanderson commented on api/blast_ui.api.inc:
```diff
@@ -179,13 +269,16 @@ function run_BLAST_tripal_job($program, $query, $database, $output_filestub, $op
+// suffix.If administrator includes the suffix, it will have to be an error, as it is for  
+// the BLAST commands themselves.
   // Strip the extension off the BLAST target
-  $database = preg_replace("/(.*)\.[pn]\w\w$/", '$1', $database);
```
See comment about about being consistent about whether to strip the extension or validate when the blast database node is created.

laceysanderson commented on includes/blast_ui.admin.inc:
```diff
@@ -156,6 +164,100 @@ function blast_ui_admin_form_validate($form, &$form_state) {
+**
+** custom method disabled the web services 
+**
+*/
```
Please follow Drupal coding standards as described here: https://www.drupal.org/docs/develop/standards/coding-standards. Most importantly follow the standards for function documention blocks: https://www.drupal.org/docs/develop/coding-standards/api-documentation-and-comment-standards including the placement and spacing of the asterisks (*) since this impacts the quality of automatically generated documentation. For example,
```
/**
 * Disables the web services.
 * This function disables the Drupal service module as well as removing the service endpoint
 * for Blast services.
 */
```
Also, there is no need to call it a custom function/method.
NOTE: I'm not going to comment on all the function doc blocks that don't match standards but would appreciate all of them fixed. Thanks!

laceysanderson commented on includes/blast_ui.admin.inc:
```diff
@@ -156,6 +164,100 @@ function blast_ui_admin_form_validate($form, &$form_state) {
+    $result = module_disable($ar,true);
+    try{
+    $eid = 1;
+    $num_deleted = db_delete('services_endpoint')->condition('eid', $eid)->execute();  
```
Here we are just deleting the first service endpoint, correct? That seems risky since some Tripal sites already use Drupal services and may have their own custom endpoints before installing blast. It would be aweful if we deleted! Perhaps it would be best if we stored the eid of the service endpoint created in the enable function in a drupal variable. Then we could variable_get() it here to ensure we only delete our own service endpoint.

prateekgta replied:
I have made changes on suggestion. I have taken care of this issue like you have suggested precisely here.

laceysanderson commented on includes/blast_ui.admin.inc:
```diff
@@ -156,6 +164,100 @@ function blast_ui_admin_form_validate($form, &$form_state) {
+  try{
+    $nid = db_insert('services_endpoint')->fields(
+    array(
+      'eid' => 1,
```
Is it valid to set this directly? What happens if there is already a service endpoint with the eid?

prateekgta replied:
I have make changes in code where it will check the max value of 'eid' and add one more to it to create the new eid in database, Now it will not conflict with the existing services installed in database.

laceysanderson commented on includes/blast_ui.admin.inc:
```diff
@@ -166,10 +268,18 @@ function blast_ui_admin_form_submit($form, $form_state) {
+  
+  if($form_state['values']['enable_blast_webservices'] == 1){
+    enable_web_services();
+  }else{
```
Coding standards: else should be on a separate line:
```
if (true) {
}
else {
}
```

prateekgta replied:
suggested changes are made.

laceysanderson commented on includes/blast_ui.linkouts.inc:
```diff
@@ -210,14 +226,12 @@ function tripal_blast_generate_linkout_gbrowse($url_prefix, $hit, $info, $option
   // Calculate the minimum & maximum coordinates.
   $min = min($coords);
   $max = max($coords);
-
```
Please do not remove my spacing. It's important to readability...

laceysanderson commented on theme/blast_help.tpl.php:
```diff
@@ -7,23 +7,233 @@
-<li><a href="<?php print url('node/add/blastdb');?>">Create "Blast Database" nodes</a> for each dataset you want to make available for your users to BLAST against. BLAST databases should first be created using the command-line <code>makeblastdb</code> program with the <code>-parse_seqids</code> flag.</li>
-<li>It's recommended that you also install the <a href="http://drupal.org/project/tripal_daemon">Tripal Job Daemon</a> to manage BLAST jobs and ensure they are run soon after being submitted by the user. Without this additional module, administrators will have to execute the tripal jobs either manually or through use of cron jobs.</li>
+  <li>
+    Install NCBI BLAST+ on your server (Tested with 2.2.26+). There is a 
```
Thanks for adding to the Help page!

laceysanderson reviewed (changes requested):
First off, Thanks for taking on this functionality! Here is  my first pass through the code without having tested it directly.

Commit:
Improved readability of webservice API parameters

Commit:
Improved spacing

Commit:
Checks BLAST target path to make sure it does not contain a file type extension.

Commit:
Removed regex's to strip off BLAST target file type extensions

Commit:
Don't add 'Select a Dataset' in get_blast_database_options() to keep it more general

Commit:
Add 'Select a Database' as an #empty parameter for form element

Commit:
Standardized handling of output files so that web services and the UI handle them the same way.

Commit:
Clarified API documentation

Commit:
Cleaned up and brought in closer alignment with Drupal coding standards

prateekgta committed:
changes in blast_ui.admin.inc in web_enable and disable module

ekcannon commented:
I think we have addressed all of the concerns. Ready for re-checking and testing.

laceysanderson commented:
Awesome, Thanks! I'll try to look at it within the next couple days

laceysanderson commented:
The changes look good --now I'll just test it on my system and we can get this merged! Sorry for the delay on this and thanks for the changes!

laceysanderson commented:
Enabling seems to work fine -- I like that it just enables the modules for me :-)

`[mysite]/restapi/blast.json` seems to work fine but I get HTTP ERROR 500 when trying `[mysite]/restapi/blast/getDatabaseOptions.json?blast_program=blastn` Looking in the log I see:

> PDOException: SQLSTATE[22P02]: Invalid text representation: 7 ERROR: invalid input syntax for integer: "getDatabaseOptions" LINE 1: SELECT job_name FROM tripal_jobs WHERE job_id = 'getDatabase... ^: SELECT job_name FROM tripal_jobs WHERE job_id = :job_id ; Array ( [:job_id] => getDatabaseOptions ) in getBlastData() (line 560 of /var/www/dev/tripal-lacey/sites/all/modules/tripal_blast/includes/blast_ui.services.inc).

Thoughts?

ekcannon commented:
Part of the problem may be because the getDatabaseOptions is expecting the program name in a POST variable. That said, it should still work as a GET.

The error is in getBlastData(), which shouldn't be called by the getDatabaseOptions command. It appears that the code attached to getDatabaseOptions isn't being called, but instead the generic code attached to [mysite]/restapi/blast is being executed. Haven't yet been able to see why.

Web services API defined in blast_ui.services.inc: blast_ui_services_resources().

Commit:
A call to getBlastData without a valid job id indicates an incorrect request

Commit:
added sample code for https

ekcannon commented:
I decided to have a GET request to getDatabaseOptions fail with a 400 error (bad request) and a message stating the likely reason: "The web service request was incorrectly specified. Check the endpoint and data, and be sure to use the correct request method (GET vs POST)"

Commit:
Improved message when one or more queries has no matches to the target

laceysanderson committed:
Merge branch 'webservices' of https://github.com/LegumeFederation/tripal_blast into LegumeFederation-webservices

laceysanderson committed:
Fixed variable max_eid missing $

laceysanderson committed:
Added Admin form to test webservices via HTTP

laceysanderson commented:
Sorry about the delay on this one. I've gone in and tested the web services via HTTP. To aid in this process I added an administrative form for testing your own web services. See code in https://github.com/tripal/tripal_blast/tree/LegumeFederation-webservices branch.

I cannot get the last service (retrieving the blast results) to work... See https://github.com/tripal/tripal_blast/blob/LegumeFederation-webservices/includes/blast_ui.services.test.inc#L231

The response I keep getting is
```
Service: Retrieve Results.
URL: http://[mysite]/restapi/blast/118.json
Results: stdClass Object
(
    [status] => 400
    [msg] => The web service request was incorrectly specified. Check the endpoint and data, and be sure to use the correct request method (GET vs POST)
);
```

I have tested the exact query sequence and database combination through the original blast interface with no issues. Furthermore, I've confirmed the job_id (118) is correct.

Thoughts?

Commit:
Fixed is_integer/intval confusion and clarified some error messages

Commit:
Fixed some bugs in sample code

Commit:
Merge branch 'LegumeFederation-webservices' of https://github.com/tripal/tripal_blast into webservices

ekcannon commented:
$job_id was of type string, so is_integer() was always false in getBlastData(). Also, the error reported was rather opaque, so I fixed that too. The code sample in blast_help.tpl.php had several errors, also fixed.

Commit:
Always produces ASN.1 output because it can be converted into any other format and because the requested format isn't being read from the URL request

ekcannon commented:
The Drupal Services module is proving to be very unstable in the BLAST UI module, so this approach to implementing web services has been abandoned.
